### DISCUSSION TITLE
Can not cognize type of model's real type in mypy

### DISCUSSION THREAD

**Discussion by barking-code** [apache/airflow#40619] [General]
### Apache Airflow version

Other Airflow 2 version (please specify below)

### If "Other Airflow 2 version" selected, which one?

2.9.0

### What happened?

This code makes `Incompatible types` error after updating to Airflow `2.7.0` from `2.6.3`.
I used `sqlalchemy2-stubs==0.0.2a38` and `mypy==1.5.0` in both versions. And I also use python3.10.14
```python
from airflow.hooks.base import BaseHook

s: str = BaseHook.get_connection("conn_id").conn_id  # raise an Incompatible types error
```

### 1. install airflow
```bash
pip install apache-airflow==2.7.0 --constraint https://raw.githubusercontent.com/apache/airflow/constraints-2.7.0/constraints-3.10.txt
```

### 2. install mypy and sqlalchemy2-stubs
```bash
pip install mypy==1.5.0 sqlalchemy2-stubs==0.0.2a38
```

### 3. copy mypy.ini
```ini
[mypy]
plugins = sqlalchemy.ext.mypy.plugin

```

### 4. copy code
```python
from airflow.hooks.base import BaseHook

s: str = BaseHook.get_connection("conn_id").conn_id  # raise an Incompatible types error
```

### 5. run mypy
```bash
mypy mypy --config-file={path of mypy.ini} {path of python file}
```

### Operating System

macOS Ventura: 13.0.1 (22A400)

### Deployment

Other

### Are you willing to submit PR?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/apache/airflow/blob/main/CODE_OF_CONDUCT.md)

**Comment by boring-cyborg** [comment]
Thanks for opening your first issue here! Be sure to follow the issue template! If you are willing to raise PR to address this issue please do so, no need to wait for approval.

**Comment by potiuk** [comment]
I think it's a very good issue - but you should likely open it in sqlalchemy repo.

**Comment by barking-code** [comment]
Hi, @potiuk. First of all, thank you for your comment :)
I think it's an Airflow problem because I used the same versions of `SQLAlchemy`, and `sqlalchemy2-stubs`.
The only major change is the Airflow Version(`2.6.2` -> `2.9.0`)

I also tested further and found that the error was raised when the Airflow version changed from `2.6.3` to `2.7.0`.
In this case, `SQLAlchemy` and `sqlalchemy2-stubs` versions were not changed.

`SQLAlchemy`: `1.4.49`
`sqlalchemy2-stubs`: `0.0.2a38`

I apologize for not being well-versed in SQLAlchemy, and I would appreciate it if you could share your opinion.

**Comment by potiuk** [comment]
Well. The community explicitly specified in our documentation that Airflow DB models ARE NOT a part of public interface of Airflow - https://airflow.apache.org/docs/apache-airflow/stable/public-airflow-interface.html#what-is-not-part-of-the-public-interface-of-apache-airflow so if you decide to use them directly, you are pretty much on your own to solve your problems. Our CI verify with mypy all the usages of sqlalchemy for Airflow, so this is fine. But if you decided to do somethin with SQLAlchemy models on your own - against the advices we gave at the "Public interface" then if something does not work for you, you won't get much help from Airflow maintainers (we specifically mentioned that so that we do not have to loose our time on solving similar issues). My recommendation is to stop modifying Airflow models (becasue they are internal detail of Airflow) and find other ways to achieve whatever you want to achieve - Airflow 3 will likely have HUGE changes to the database, including likely rewriting some of the base models, so you should not rely on thos models to continue working as they are.

If you chose to work with SQL Alchemy models directly, you need to become expert in that - this is noit something you will get a lot of help here, this is why I suggested asking this question in SQLAlachemy forums.

**Comment by barking-code** [comment]
@potiuk 
Thank you for your kind explanation. 
I think I caused a misunderstanding. I am not modifying Airflow DB model. The code I shared is test code for raising the same error I encountered. I am just using it this way.
(e.g `id: str = BaseHook.get_connection().conn_id`)
So, I expected a bit of help from the community. You mentioned that all the usages of SQLAlchemy for Airflow were verified with mypy, so I think something is wrong with my local setup.
Thank you for your opinion and for the notice about the changes in Airflow 3.

**Reply by potiuk** [reply]
So maybe better to explain the exact error you have and intentions you have, rather than coming up with generic SQL-alchemy-looking-like piece of code? I am sure that if you explain exactly what you try to do (from these explanations you gave I am still not sure what kind of error you have - because the original example and what you shared in your comment are pretty different).

So you might get help, but make sure you specify the problem you have in the way that will allow others to help you,

**Reply by barking-code** [reply]
@potiuk 
I updated the issue to make it simpler. I found the code causing the problem, so I will open a PR and fix it.
I appreciate your interest. :)
